packages/coinjoin/src/client/CoinjoinRound.ts

```diff
@@ -173,15 +156,20 @@ export class CoinjoinRound extends EventEmitter implements SerializedCoinjoinRou
         return this;
     }
 
-    async process(accounts: Account[]) {
+    async process(accounts: Account[], prison: CoinjoinPrison) {
         const { log } = this.options;
         if (this.inputs.length === 0) {
             log('Trying to process round without inputs');
             return this;
         }
-        await this.processPhase(accounts);
+        await this.processPhase(accounts, prison);
 
-        const [inputs, failed] = arrayPartition(this.inputs, input => !input.error);
+        const [inputs, failed] = arrayPartition(this.inputs, input => {
+            if (input.error) {
+                prison.detain(input.outpoint, { roundId: this.id, reason: input.error.message });
+            }
+            return !input.error;
+        });
```

Comment:
I'd prefer not to have side effects inside selector function, and it'll be shorter as well.
```suggestion
        const [inputs, failed] = arrayPartition(this.inputs, input => !input.error);
        failed.forEach(input =>
            prison.detain(input.outpoint, { roundId: this.id, reason: input.error?.message }),
        );
```

Reply:
[732481b](https://github.com/trezor/trezor-suite/pull/6712/commits/732481b70daf5567bd5b23c089c3b2a292284be5)